peachjam/resources.py

```diff
@@ -38,54 +38,93 @@
+
     def clean(self, value, row=None, *args, **kwargs):
-        if value:
-            return self.model.objects.get(code__iexact=value)
-        return None
+        if not value:
+            raise ValidationError(f"{self.name} is required")
+        return self.model.objects.get(**{self.field: value})
 
 
-class SourceFileWidget(CharWidget):
+class SourceFileWidget(CharRequiredWidget):
     def clean(self, value, row=None, **kwargs):
+        super().clean(value)
+        source_url = self.get_source_url(value)
         try:
-            r = requests.head(value)
+            r = requests.head(source_url)
             r.raise_for_status()
-            return value
-        except requests.exceptions.HTTPError as e:
+            return source_url
+        except requests.exceptions.RequestException as e:
             raise ValidationError(e)
 
+    @staticmethod
+    def get_source_url(value):
+        source_url = None
+        source_files = [x.strip() for x in value.split("|")]
+        docx = re.compile(r"\.docx?$")
+        pdf = re.compile(r"\.pdf$")
+        for file in source_files:
+            match_docx = docx.search(file)
+            match_pdf = pdf.search(file)
+            # prefer the .docx file if available, otherwise use .pdf, ignore .rtf
+            if match_docx:
+                source_url = file
+                break
+            elif match_pdf:
+                source_url = file
+
+        return source_url
+
 
 class SkipRowWidget(BooleanWidget):
     def clean(self, value, row=None, **kwargs):
         return bool(value)
 
 
+class AuthorWidget(ForeignKeyWidget):
+    def clean(self, value, row=None, *args, **kwargs):
+        if not value:
+            raise ValidationError("author is required")
+        return self.model.objects.get(code__iexact=value)
+
+
 class BaseDocumentResource(resources.ModelResource):
     author = fields.Field(
-        column_name="author",
         attribute="author",
         widget=AuthorWidget(Author),
     )
     language = fields.Field(
         attribute="language",
-        widget=ForeignKeyWidget(Language, field="iso_639_2T"),
+        widget=ForeignKeyRequiredWidget(Language, name="language", field="iso_639_2T"),
     )
     jurisdiction = fields.Field(
-        attribute="jurisdiction", widget=JurisdictionWidget(Country)
+        attribute="jurisdiction",
+        widget=ForeignKeyRequiredWidget(Country, name="jurisdiction", field="iso"),
     )
     locality = fields.Field(
-        attribute="locality", widget=ForeignKeyWidget(Locality, field="code")
+        attribute="locality",
+        widget=ForeignKeyRequiredWidget(Locality, name="locality", field="code"),
+    )
+    source_url = fields.Field(
+        attribute="source_url", widget=SourceFileWidget(field="source_url")
     )
-    source_url = fields.Field(attribute="source_url", widget=SourceFileWidget())
     skip = fields.Field(attribute="skip", widget=SkipRowWidget())
 
+    required_fields = None

```

Comment:
`set(None)` raises an exception, so perhaps set this to `[]`

Reply:
Changed. Thanks